### ISSUE TITLE
Cannot find module 'expo-modules-autolinking'

### ISSUE THREAD

**Issue by tkrajina** [expo/expo#13580]
### Summary

When building the app it fails with the following error:

```
Cannot find module 'expo-modules-autolinking'
```

In my current application I followed all the steps given by @byCedric in https://github.com/expo/expo/issues/13475#issuecomment-877296024 without success.

I also tried on a bare-new (bare) app. 

react-native-adapter version:

```
$ grep react-native-adapter yarn.lock -A 1                                                                                                                    main *
"@unimodules/react-native-adapter@~6.3.3", "@unimodules/react-native-adapter@~6.3.4":
  version "6.3.4"
```

Node version

```
$ node --version
v14.17.0
```

OS: BigSur on a Macbookpro with M1

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

bare

### What platform(s) does this occur on?

Android

### Environment

```
$ expo diagnostics                                                                                                                                            main *

  Expo CLI 4.7.3 environment info:
    System:
      OS: macOS 11.4
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 14.17.0 - /opt/homebrew/bin/node
      Yarn: 1.22.10 - /opt/homebrew/bin/yarn
      npm: 6.14.13 - /opt/homebrew/bin/npm
      Watchman: 4.9.0 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.10.1 - /usr/local/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 14.3, DriverKit 20.2, macOS 11.1, tvOS 14.3, watchOS 7.2
    IDEs:
      Android Studio: EAP AI-203.7717.56.2031.7479365 AI-203.7717.56.2031.7479365
      Xcode: 12.3/12C33 - /usr/bin/xcodebuild
    npmPackages:
      expo: ~42.0.1 => 42.0.1 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native: ~0.63.4 => 0.63.4 
      react-native-web: ~0.13.12 => 0.13.18 
    npmGlobalPackages:
      expo-cli: 4.7.3
    Expo Workflow: bare
```

### Reproducible demo or steps to reproduce from a blank project

Create an initial app with:

```
tkr ~/projects/tmp $ expo init                                                                                                                                          ispravljanje-originalne-verzije *
✔ What would you like to name your app? … expo-playground-2
✔ Choose a template: › blank (TypeScript)  same as blank but with TypeScript configuration
✔ Downloaded and extracted project files.
🧶 Using Yarn to install packages. Pass --npm to use npm instead.
✔ Installed JavaScript dependencies.

✅ Your project is ready!

To run your project, navigate to the directory and run one of the following yarn commands.

- cd expo-playground-2
- yarn start # you can open iOS, Android, or web from here, or run them directly with the commands below.
- yarn android
- yarn ios
- yarn web
```

Then eject:

```
tkr ~/projects/tmp $ cd expo-playground-2                                                                                                                               tkr ~/projects/tmp/expo-playground-2 $ expo eject                                                                                                                                                    Your git working tree is clean
To revert the changes after this command completes, you can run the following:
  git clean --force && git reset --hard

📝  Android package Learn more: https://expo.fyi/android-package

✔ What would you like your Android package name to be? … com.tkrajina.expoplayground2
📝  iOS Bundle Identifier Learn more: https://expo.fyi/bundle-identifier

✔ What would you like your iOS bundle identifier to be? … com.tkrajina.expoplayground2
✔ Created native projects
✔ Added Metro config.
✔ Updated package.json and added index.js entry point for iOS and Android.
› Removed "main": "node_modules/expo/AppEntry.js" from package.json because we recommend using index.js as main instead.

🧶 Using Yarn to install packages. Pass --npm to use npm instead.
✔ Cleaned JavaScript dependencies.
✔ Installed JavaScript dependencies.
⠙ Config syncing
Using node to generate images. This is much slower than using native packages.
› Optionally you can stop the process and try again after successfully running `npm install -g sharp-cli`.

✔ Config synced
✔ Installed pods and initialized Xcode workspace.

➡️  Next steps
› 💡 You may want to run npx @react-native-community/cli doctor to help install any tools that your app may need to run your native projects.
› 🔑 Download your Android keystore (if you're not sure if you need to, just run the command and see): expo fetch:android:keystore
› 📁 The property assetBundlePatterns does not have the same effect in the bare workflow.
  Learn more: https://docs.expo.io/bare/updating-your-app/#embedding-assets

☑️  When you are ready to run your project
To compile and run your project in development, execute one of the following commands:
› yarn ios
› yarn android
› yarn web
```

Then, when building it from android studio:

<img width="957" alt="Screenshot 2021-07-12 at 09 32 39" src="https://user-images.githubusercontent.com/379243/125248281-2c6a6580-e2f4-11eb-8c0b-d59ff8d1eada.png">

**Comment by tkrajina** [comment]
After some additional testing, I found that the command `exec()` is running is:

```
node --eval require('expo-modules-autolinking')(process.argv.slice(1)) -- generate-package-list --platform android --namespace org.unimodules.adapters.react --target /Users/tkr/projects/coachbit/app-sdk-42/node_modules/@unimodules/react-native-adapter/android/build/generated/expo/src/main/java/org/unimodules/adapters/react/ExpoModulesPackageList.java --empty
```

...and that fails. It works only after I put `require('expo-modules-autolinking')(process.argv.slice(1))` in additional double quotes:

```
node --eval "require('expo-modules-autolinking')(process.argv.slice(1))" -- generate-package-list --platform android --namespace org.unimodules.adapters.react --target /Users/tkr/projects/coachbit/app-sdk-42/node_modules/@unimodules/react-native-adapter/android/build/generated/expo/src/main/java/org/unimodules/adapters/react/ExpoModulesPackageList.java --empty
```

...and that works.

For that reason I changed `convertOptionsToCommandArgs` to add quotes in the require parameter:

```
  static private String[] convertOptionsToCommandArgs(String command, Map options) {
    String[] args = [
      'node',
      '--eval',
      '"require(\'expo-modules-autolinking\')(process.argv.slice(1))"',
      '--',
      command,
      '--platform',
      'android'
    ]
```

...aaand this solved my problem in the bare app I created for this. In my production app there is another bug happening later but it's probably unrelated (I'll investigate further).

**Comment by byCedric** [comment]
Hi @tkrajina, thanks for the issue report. Can you share a bit more about your project and/or environment? Is it a monorepo, or just a normal project? And do you have anything special installed, e.g. Node through a versions manager? 

I created this repository to validate the behavior of macOS, Linux, and Windows with this module. Unfortunately, [it seems to run fine](https://github.com/byCedric/expo-issue-13475/actions/runs/1022656261) ([even when ejecting too](https://github.com/byCedric/expo-issue-13475/actions/runs/1022802754)). If we can reproduce this issue properly, we can work on a fix.

**Comment by tkrajina** [comment]
@byCedric I can detereministically reproduce that problem by just "expo init" and "expo eject". And it always ends up with this problem. And it happends in that newly created project and in my existing project. 

Node is installed via homebrew, nothing fancy. In fact this machine is a pretty new macbook pro with everything freshly installed via homebrew.

About the error "can't find module" -- it's misleading. it will be thrown even when `expo-modules-autolinking` **is** installed but `exec()` fails running the command. For example, try removing one parentheses from `require(\'expo-modules-autolinking\')(process.argv.slice(1))` and it will still fail saying that it can't find the `expo-modules-autolinking`. It's not a bug, but the error message should tell that a specific command failed to run, not that the dependency is missing. The dependency *is* there but running `node --eval ...` failed.

That said, I can also reproduce it by just running that node command in the shell. The command run by `exec()` is:

```
$ node --eval require('expo-modules-autolinking')(process.argv.slice(1)) -- generate-package-list --platform android --namespace org.unimodules.adapters.react --target /Users/tkr/projects/tmp/expo-playground-2/node_modules/@unimodules/react-native-adapter/android/build/generated/expo/src/main/java/org/unimodules/adapters/react/ExpoModulesPackageList.java --empty
```

..and it fails with:

```
zsh: no matches found: require(expo-modules-autolinking)(process.argv.slice(1))
```

with quotes it runs without problems:

```
$ node --eval "require('expo-modules-autolinking')(process.argv.slice(1))" -- generate-package-list --platform android --namespace org.unimodules.adapters.react --target /Users/tkr/projects/tmp/expo-playground-2/node_modules/@unimodules/react-native-adapter/android/build/generated/expo/src/main/java/org/unimodules/adapters/react/ExpoModulesPackageList.java --empty
```

Also, to be sure it's not only a zsh problem, I tried it with bash:

With quotes runs fine:

```
bash-3.2$ node --eval "require('expo-modules-autolinking')(process.argv.slice(1))" -- generate-package-list --platform android --namespace org.unimodules.adapters.react --target /Users/tkr/projects/tmp/expo-playground-2/node_modules/@unimodules/react-native-adapter/android/build/generated/expo/src/main/java/org/unimodules/adapters/react/ExpoModulesPackageList.java --empty
```

But without quotes:

```
bash-3.2$ node --eval require('expo-modules-autolinking')(process.argv.slice(1)) -- generate-package-list --platform android --namespace org.unimodules.adapters.react --target /Users/tkr/projects/tmp/expo-playground-2/node_modules/@unimodules/react-native-adapter/android/build/generated/expo/src/main/java/org/unimodules/adapters/react/ExpoModulesPackageList.java --empty
bash: syntax error near unexpected token `('
```

Now, that's shell. But I suspect there is something similar with the environment which runs `exec()` that makes it fail without quotes (and results in a misleading message saying that the problem is a missing `expo-modules-autolinking`).

**Comment by byCedric** [comment]
I agree that it's not a great error, and pretty hard to debug. The commands you listed are also failing, as expected because `eval` requires a string of js code.

But your [change failed in our CI](https://github.com/expo/expo/pull/13581/checks?check_run_id=3045148741#step:10:24). This leads me to believe something else is going on because the current one actually passes. This change also crashes my local Android builds (macOS), which leads me to believe something else is going on.

I can't repro the exact same scenario where it fails for you. If you know more about this issue and make the CI pass in any change, I'd be happy to review it and test it on Windows, Mac, and Linux.

**Comment by tkrajina** [comment]
👍 I'll try testing it on a few other machines and let you know.

**Comment by owaiswiz** [comment]
I was facing a similar issue running `expo run:android`. I upgraded gradle to 6.9 from 6.8 and that fixed it. Also on M1 Big Sur.

**Comment by sohcah** [comment]
Same issue here on M1 / Big Sur.

**Comment by RodolfoGS** [comment]
I have the same issue after updating **Android Studio** to the latest version (`2020.3.1`). Also on M1 Big Sur.

**Comment by tsapeta** [comment]
Yes, I confirm this issue may happen on M1 devices with Gradle older than 6.9 — which is the first version with native support for Apple Silicon so it's worth upgrading either way, just like I did in this repo here #13929

**Comment by endruuu** [comment]
Thank you so much! Upgrading gradle to 6.9 fixes the issue.

[This issue](https://github.com/expo/expo/issues/13475) might also be related to this.
Since I can't comment anymore there, I comment here just in case someone faces that issue.

**Comment by apple-blossom** [comment]
I fixed similar error this way - installed expo-modules-autolinking
then ran `npx expo install --fix`

**Comment by burakgokcinarr** [comment]
Hello,

I was getting the following error in the Android part with Expo prebuild.

![Ekran Resmi 2023-08-18 11 03 24](https://github.com/expo/expo/assets/18294228/12c0a924-09ec-461b-a04d-019dc83af445)

This is my Expo app.json file content.

![Ekran Resmi 2023-08-18 11 05 45](https://github.com/expo/expo/assets/18294228/ed4661d2-b96a-4947-8a8b-2b5c5829aec4)

I guess the problem is the "expo prebuild" command is using 8.0.1 as the gradle version by default. And the default kotlin version included in the build.gradle file is 1.8.10. The incompatibility starts here. Doing the following procedure fixed the problem for me. I changed the Kotlin version to "1.9.0"

build.gradle 

I made the following change in the build.gradle file

kotlinVersion = findProperty('android.kotlinVersion') ?: '1.8.10' 
change
kotlinVersion = '1.9.0'

Then the project run " sync project width gradle file "

hope it helps :)
